packages/driver/src/cy/net-stubbing/events/index.ts

```diff
@@ -0,0 +1,69 @@
+import { Route, Request } from '../types'
+import { NetEventFrames } from '@packages/net-stubbing/lib/types'
+import { onRequestReceived } from './request-received'
+import { onResponseReceived } from './response-received'
+import { onRequestComplete } from './request-complete'
+import Bluebird from 'bluebird'
+
+export type HandlerFn<Frame extends NetEventFrames.BaseHttp> = (Cypress: Cypress.Cypress, frame: Frame, opts: {
+  getRequest: (routeHandlerId: string, requestId: string) => Request | undefined
+  getRoute: (routeHandlerId: string) => Route | undefined
+  emitNetEvent: (eventName: string, frame: any) => Promise<void>
+  failCurrentTest: (err: Error) => void
+}) => Promise<void> | void
+
+const netEventHandlers: { [eventName: string]: HandlerFn<any> } = {
+  'http:request:received': onRequestReceived,
+  'http:response:received': onResponseReceived,
+  'http:request:complete': onRequestComplete,
+}
+
+export function registerEvents (Cypress: Cypress.Cypress) {
+  const { state } = Cypress
+
+  function getRoute (routeHandlerId) {
+    return state('routes')[routeHandlerId]
+  }
+
+  function getRequest (routeHandlerId: string, requestId: string): Request | undefined {
+    const route = getRoute(routeHandlerId)
+
+    if (route) {
+      return route.requests[requestId]
+    }
+
+    return
+  }
+
+  function emitNetEvent (eventName: string, frame: any): Promise<void> {
+    // all messages from driver to server are wrapped in backend:request
+    return Cypress.backend('net', eventName, frame)
+  }
+
+  function failCurrentTest (err: Error) {
+    // @ts-ignore
+    // NOTE: this only works to fail the test if a cy chain is active
+    Cypress.action('cy:fail', err, state('runnable'))
```

Comment:
I worry about the need to do this, since it's bypassing all the logic where we send `cy:fail` here: https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cypress/cy.js#L801. We want all errors to go through that `fail` function so they get the full benefit of the error improvements.

Reply:
perfect, `cy.fail` did not exist at the time this PR was written, happy to use it now

do you know if `cy.fail` will work around the bug #8221 and fail the test, even if no cy command is active?

Reply:
I'm afraid `cy.fail` doesn't exist. I wish it were that simple. I think we might need a more creative solution for this to ensure it makes it into that `fail` function. Let me know if you'd like to pair on it.

Reply:
fixed this, also, cleaned up the indentation/nesting of error messages in the driver